### PR TITLE
move timeouts and cancellation handling to remote_storage

### PR DESCRIPTION
**Pull request by koivunej**
Cancellation and timeouts are handled at remote_storage callsites, if they are. However they should always be handled, because we've had transient problems with remote storage connections.

- Add cancellation token to the `trait RemoteStorage` methods
    - For `download*`, `list*` methods there is `DownloadError::{Cancelled,Timeout}`
    - For the rest now using `anyhow::Error`, it will have root cause `remote_storage::TimeoutOrCancel::{Cancel,Timeout}`
    - Both types have `::is_permanent` equivalent which should be passed to `backoff::retry`
- New generic RemoteStorageConfig option `timeout`, defaults to 120s
- Start counting timeouts only after acquiring concurrency limiter permit
- Cancellable permit acquiring
- Download stream timeout or cancellation is communicated via an `std::io::Error`
- Exit backoff::retry by marking cancellation errors permanent

Fixes: #6096
Closes: #4781

### CODE REVIEW

File: libs/remote_storage/src/azure_blob.rs [hunk 1]
```diff
@@ -130,32 +131,42 @@ impl AzureBlobStorage {
...
             }
-            let data = part
-                .data
-                .collect()
-                .await
-                .map_err(|e| DownloadError::Other(e.into()))?;
-            bufs.push(data);
+            Ok(Download {
+                // thsi is actually 'static, but we cannot mention it in the signature, but it can be
+                // substituted for the 'a on &'a CancellationToken
```
Comment (koivunej) [inline libs/remote_storage/src/azure_blob.rs:157]:
```suggestion
```

File: libs/remote_storage/src/azure_blob.rs [hunk 2]
```diff
@@ -295,82 +347,123 @@ impl RemoteStorage for AzureBlobStorage {
...
+                };
+                copy_status = Some(status);
             }
-            copy_status = status;
+        };
+
+        tokio::select! {
+            res = op => res,
+            _ = cancel.cancelled() => Err(anyhow::Error::new(TimeoutOrCancel::Cancel)),
+            _ = timeout => Err(anyhow::Error::new(TimeoutOrCancel::Timeout).context(format!("Timeout, last state: {copy_status:?}"))),
```
Comment (koivunej) [inline libs/remote_storage/src/azure_blob.rs:466]:
```suggestion
            _ = timeout => Err(anyhow::Error::new(TimeoutOrCancel::Timeout).context(format!("Timeout, last status: {copy_status:?}"))),
```

File: safekeeper/src/wal_backup.rs [hunk 3]
```diff
@@ -510,7 +510,13 @@ async fn backup_object(
 
     let file = tokio_util::io::ReaderStream::with_capacity(file, BUFFER_SIZE);
 
-    storage.upload_storage_object(file, size, target_file).await
+    let cancel = CancellationToken::new();
+    let timeout = std::time::Duration::from_secs(120);
+
+    // TODO: is there upper retry?
```
Comment (koivunej) [inline safekeeper/src/wal_backup.rs:516]: TODO or follow-up

File: safekeeper/src/wal_backup.rs [hunk 4]
```diff
@@ -594,7 +609,13 @@ pub async fn copy_s3_segments(
 
     let remote_path = RemotePath::new(&relative_dst_path)?;
 
-    let files = storage.list_files(Some(&remote_path)).await?;
+    let timeout = Duration::from_secs(120);
+    let cancel = CancellationToken::new();
+
+    // FIXME: is there upper retry?
```
Comment (koivunej) [inline safekeeper/src/wal_backup.rs:615]: TODO or follow-up.

File: libs/remote_storage/src/azure_blob.rs [hunk 5]
```diff
@@ -303,20 +315,27 @@ impl RemoteStorage for AzureBlobStorage {
...
     }
 
-    async fn download(&self, from: &RemotePath) -> Result<Download, DownloadError> {
+    async fn download(
+        &self,
+        from: &RemotePath,
+        timeout: Duration,
+        cancel: &CancellationToken,
+    ) -> Result<Download, DownloadError> {
         let _permit = self.permit(RequestKind::Get).await;
```
Comment (koivunej) [inline libs/remote_storage/src/azure_blob.rs:322]: Move permit to the download_for_builder

File: pageserver/src/tenant.rs [hunk 6]
```diff
@@ -3405,8 +3405,11 @@ impl Tenant {
                 );
                 let dest_path =
                     &remote_initdb_archive_path(&self.tenant_shard_id.tenant_id, &timeline_id);
+
+                // FIXME: backoff::retry
+                // FIXME: this seems out of place
```
Comment (koivunej) [inline pageserver/src/tenant.rs:3410]: Fixme

File: libs/remote_storage/src/azure_blob.rs (hunk 1 above)
Comment (koivunej) [reply libs/remote_storage/src/azure_blob.rs:157]: 656a12240

File: libs/remote_storage/src/azure_blob.rs (hunk 5 above)
Comment (koivunej) [reply libs/remote_storage/src/azure_blob.rs:322]: 7f34e43c8

File: libs/remote_storage/src/s3_bucket.rs [hunk 7]
```diff
@@ -267,6 +284,9 @@ impl S3Bucket {
         let body = PermitCarrying::new(permit, body);
         let body = TimedDownload::new(started_at, body);
 
+        let cancel_or_timeout = crate::support::cancel_or_timeout(remaining, cancel.clone());
+        let body = crate::support::DownloadStream::new(cancel_or_timeout, body);
+
         Ok(Download {
             metadata,
             etag,
```
Comment (koivunej) [inline libs/remote_storage/src/s3_bucket.rs:292]: Can't say any of these variations would look really great to me. I was initially hoping to do a layered approach but with the method parameters it doesn't at least feel natural; basically one would had had just one type handling the concurrency limiting and timing out with the necessary complication.

I guess we are left with `acquire_permit_then_timeout(async { ... })` kind of solution.

File: libs/remote_storage/src/lib.rs [hunk 8]
```diff
@@ -152,9 +160,11 @@ pub trait RemoteStorage: Send + Sync + 'static {
     async fn list_prefixes(
         &self,
         prefix: Option<&RemotePath>,
+        timeout: Duration,
```
Comment (arpad-m) [inline libs/remote_storage/src/lib.rs:163]: I don't think that for a remote storage API one needs per-call specific timeout settings.

I think the timeout should be a global setting that is part of the struct and that can be edited via a `set_timeout` function, that's maybe part of the trait.

Comment (koivunej) [reply libs/remote_storage/src/lib.rs:163]: Agreed this was a waste putting it on every operation. I'll just make that configurable via config.

File: libs/remote_storage/src/s3_bucket.rs (hunk 7 above)
Comment (koivunej) [reply libs/remote_storage/src/s3_bucket.rs:303]: I moved the cancel to permit acquiring as it used to be in my original PR.

File: libs/remote_storage/src/lib.rs (hunk 8 above)
Comment (koivunej) [reply libs/remote_storage/src/lib.rs]: There is now a general configuration parameter for remote storage which is copied into per adapter configuration. There is no mutability.

File: libs/remote_storage/src/azure_blob.rs [hunk 9]
```diff
@@ -309,82 +364,115 @@ impl RemoteStorage for AzureBlobStorage {
...
+            // https://github.com/Azure/azure-sdk-for-rust/issues/1249
+            for path in paths {
+                let blob_client = self.client.blob_client(self.relative_path_to_name(path));
+
+                let request = blob_client.delete().into_future();
+
+                let res = tokio::select! {
+                    res = request => res,
+                    _ = tokio::time::sleep(self.timeout) => return Err(TimeoutOrCancel::Timeout.into()),
+                };
```
Comment (koivunej) [inline libs/remote_storage/src/azure_blob.rs:394]: Could be a timeout

File: libs/remote_storage/src/support.rs [hunk 10]
```diff
@@ -31,3 +37,222 @@ impl<S: Stream> Stream for PermitCarrying<S> {
...
+impl From<TimeoutOrCancel> for DownloadError {
+    fn from(value: TimeoutOrCancel) -> Self {
+        use TimeoutOrCancel::*;
+
+        match value {
+            Timeout => DownloadError::Timeout,
+            Cancel => DownloadError::Cancelled,
+        }
+    }
+}
```
Comment (koivunej) [inline libs/remote_storage/src/support.rs:176]: I was thinking of moving all of the errors to `error.rs` even though I dislike the refactoring -- would prefer to have per-operation split not per file kind..

File: libs/remote_storage/src/azure_blob.rs [hunk 11]
```diff
@@ -192,66 +214,82 @@ impl RemoteStorage for AzureBlobStorage {
...
-            for key in blob_iter {
-                res.keys.push(key);
-                if let Some(mut mk) = max_keys {
-                    assert!(mk > 0);
-                    mk -= 1;
-                    if mk == 0 {
-                        return Ok(res); // limit reached
+            let mut response = builder.into_stream();
+            let mut res = Listing::default();
+            let mut timeout = std::pin::pin!(tokio::time::sleep(self.timeout));
```
Comment (koivunej) [inline libs/remote_storage/src/azure_blob.rs:253]: This should be a per request as well.

File: libs/remote_storage/src/support.rs (hunk 10 above)
Comment (koivunej) [reply libs/remote_storage/src/support.rs]: [e47e937](https://github.com/neondatabase/neon/pull/6697/commits/e47e937b1cf9f25141436c35a939b02d0f5aed6b)

File: libs/remote_storage/src/error.rs [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+
+impl std::error::Error for DownloadError {}
+
+impl DownloadError {
+    /// Returns true if the error should not be retried with backoff
+    pub fn is_permanent(&self) -> bool {
+        use DownloadError::*;
+        match self {
+            BadInput(_) => true,
+            NotFound => true,
```
Comment (arpad-m) [inline libs/remote_storage/src/error.rs:41]:
```suggestion
            BadInput(_) | NotFound| Cancelled => true,
```

File: libs/remote_storage/src/lib.rs [hunk 13]
```diff
@@ -206,27 +215,61 @@ pub trait RemoteStorage: Send + Sync + 'static {
         data_size_bytes: usize,
         to: &RemotePath,
         metadata: Option<StorageMetadata>,
+        cancel: &CancellationToken,
     ) -> anyhow::Result<()>;
 
-    /// Streams the remote storage entry contents into the buffered writer given, returns the filled writer.
+    /// Streams the remote storage entry contents into the buffered writer given.
```
Comment (arpad-m) [inline libs/remote_storage/src/lib.rs:221]:
```suggestion
    /// Streams the remote storage entry contents into the given buffered writer.
```

Comment (koivunej) [reply libs/remote_storage/src/lib.rs:221]: 736437439

File: libs/remote_storage/src/error.rs (hunk 12 above)
Comment (koivunej) [reply libs/remote_storage/src/error.rs:41]: 6739aa42bff60d131a710fa598f265702207af79

File: libs/remote_storage/src/s3_bucket.rs [hunk 14]
```diff
@@ -278,33 +307,57 @@ impl S3Bucket {
 
     async fn delete_oids(
```
Comment (arpad-m) [inline libs/remote_storage/src/s3_bucket.rs:308]: I think having both `delete_oids` and `delete_oids0` is redundant. The `delete_oids` function is only called once. The content of `delete_oids` should IMO be moved back into `delete_objects`, and `delete_oids0` should be renamed to `delete_oids`.

File: libs/remote_storage/src/error.rs [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+            TimeTravelError::Other(e) => write!(f, "Failed to time travel recover a prefix: {e:?}"),
+        }
+    }
+}
+
+impl std::error::Error for TimeTravelError {}
+
+/// Root cause for cancellations however this type does not implement `std::error::Error` so it
+/// cannot be put as the root cause of `anyhow::Error`. It should never need to be exposed out of
+/// this crate.
```
Comment (arpad-m) [inline libs/remote_storage/src/error.rs:90]: This should explain why there is two separate structs `Cancelled` and `TimeoutOrCancel` and what their difference is: from reading the code, it seems that former implements more conversions while latter also includes timeout errors. My first reaction was to suggest unifying the two. Ideally we'd write why we didn't do that.

**Review by arpad-m — APPROVED** [review]
LGTM except for nits. Also, I suppose we don't want `pageserver/test_data/sk_wal_segment_from_pgbench.zst` :)

File: libs/remote_storage/src/s3_bucket.rs (hunk 14 above)
Comment (koivunej) [reply libs/remote_storage/src/s3_bucket.rs:308]: I think I needed to do some permit wrestling, seeing about this soon.

**Comment by koivunej** [comment]
> Also, I suppose we don't want pageserver/test_data/sk_wal_segment_from_pgbench.zst :)

Weird, I thought it didn't come from me. Why do we have this file then.

File: libs/remote_storage/src/s3_bucket.rs (hunk 14 above)
Comment (koivunej) [reply libs/remote_storage/src/s3_bucket.rs:309]: Ok, cleaned up this in [a2eeb13](https://github.com/neondatabase/neon/pull/6697/commits/a2eeb13bf55e3f3cb3b8d6b46523f6473858bf56)

**Comment by koivunej** [comment]
Latest force push got rid of the extra `zst` file I had accidentially added and missed in review mode.

File: libs/remote_storage/src/error.rs (hunk 15 above)
Comment (arpad-m) [reply libs/remote_storage/src/error.rs:87]: I think unifying is not a good idea, but it would be good to justify/explain that choice in the comment.

Comment (koivunej) [reply libs/remote_storage/src/error.rs]: Attempted that in 31ade7fa9

**Pull request merged**